### ISSUE TITLE
Make the audit capable of being run regularly

### ISSUE THREAD

**Issue by 0xdabbad00** [duo-labs/cloudmapper#483]
CloudMapper's `audit` command has a lot of checks in it, and it would be great to run this nightly in environments and send the alerts to a Slack or other output.

CloudMapper was made for doing one time assessments though.  You could today just run the following:
```
rm -rf ./account-data
python cloudmapper.py collect --account prod
python cloudmapper.py audit --account prod --json | ./send_alerts
```
where `send_alerts` would send messages to Slack (or could just email the whole blob to you).

In order to make CloudMapper work better for this use case, I'm going to do the following:
- Ensure CloudMapper has solid error-checking/handling so if something breaks in the `collect` or `audit` phase that these errors aren't swallowed.
- Identify or create a tool for `send_alerts`, possibly changing the json format of the audit output as needed.
- Add an ability to filter/mute alerts for certain things and/or cache alerts.  For example, you might have accepted that an IAM user's access keys are never going to be rolled, so you don't want an alert every night about that same user have old access keys.

**Comment by 0xdabbad00** [comment]
There is interest in making this run serverlessly.  CloudMapper's collect phase can take over an hour, so that may be problematic, but as a step for this, I'll be having the `./account-data` optionally be hosted on S3.  In theory, the collect phase could be broken up so a different lambda for each service could be run simultaneously.  A problem with that is how will the `audit` step know when to run?  I could just kick off all the collection at one point, and then 2 hours later run the `audit`, under the assumption that everything must have run.

**Comment by schosterbarak** [comment]
how about running cloudmapper in a container (since it's a long living task) and schedule it using airflow?

**Comment by jchrisfarris** [comment]
This is where you might want to use stepfunctions. When I had to solve this, the first lambda in the SF was to check to see if another SF execution was running.

**Comment by 0xdabbad00** [comment]
My current client wants this as a Lambda, as management is easier than a container.

Regarding the use of S3, I have the following options:
- Use a FUSE filesystem of some sort to keep things as they are in CloudMapper so file reads/writes get proxied to S3 transparently.  Could use Netflix's MezzFS maybe (I've done no research).
- Abstract all of the file access to go through something that can proxy calls for me from the Python code, so the python code is making S3 calls.
- Run CloudMapper pretty much as is, but have a before and after task that copies `account-data` to/from S3 and the local filesystem.

**Comment by 0xdabbad00** [comment]
For piping to Slack, there are solutions out there such as https://github.com/clarkie/pipe-to-slack
But I'll probably write my own to be able to make the messages easier to read.

An example output from `python cloudmapper.py audit --account demo --json` is:

```
{"account_id": "123456789012", "account_name": "demo", "region": "us-east-1", "issue": "S3_PUBLIC_POLICY_GETOBJECT_ONLY", "resource": "cloudmapper_demo", "details": null}
```

This would be better as:
"""
**AWS audit alert**
Severity: Info
Account: demo (123456789012) - us-east-1
Issue: Internet accessible S3 bucket via policy (only GetObject)
Resource: cloudmapper_demo
Description: This is the right way to make an S3 bucket public when you don't want to put CloudFront in front of it. This may be done when a third-party caching service is being used and you don't care about direct access to the S3 bucket.
"""
That additional info can be pulled from the `audit_config.yaml` file, as that is used in generating the report.

**Comment by jchrisfarris** [comment]
I do a lot with spreadsheet generation in Lambda. I copy the file to /tmp,
do stuff, push back to S3.

S3 FUSE is evil.

**Comment by 0xdabbad00** [comment]
Another thing that will take work is making this runnable from a Lambda, which means ensuring this can run in AWS's linux environment, so I'm going to make that work.

**Comment by FernandoMiguel** [comment]
There's a tool that converts any Docker to lambda.
But I would still prefer for a long lived task to be a container rather than serverless.
Maybe fargate?
It's easy to template with csk

**Comment by 0xdabbad00** [comment]
Some considerations for Lambda vs Container:
- I'm doing this task via contract work, and the client did request this be Lambdas. However, I don't think they are too strict on that if it's just a bad idea.
- Having it be a container is likely way easier on me (and the client), as making this a Lambda is going to involve having to break up the collection in different ways, probably tracking state of how far the collect has gone, then get another lambda invoked to pick up where the previous one left off, etc.  That would be a mess.

I think it will need to be a container, and if not, at least much of the groundwork carries over to making it a lambda in the future if that's the route needed.

**Comment by 0xdabbad00** [comment]
And as for making this run regularly.  My assumption is this will be run nightly, for maybe an hour, so I can use a scheduled CloudWatch Event trigger against Fargate. https://aws.amazon.com/about-aws/whats-new/2018/08/aws-fargate-now-supports-time-and-event-based-task-scheduling/

This is going to be auditing many accounts, so there is the architecture question of do I have a single container collect + audit, collect + audit, collect + audit.  Or do I assume a container does a single collect + audit against a single account, and that potentially hundreds of containers are spun up.  I believe the latter option of one container per account is preferred.

With containers, I still plan on using S3 to record the data because for this environment there are other tools that may be using CloudMapper's collected data so I need to make that accessible outside of the container.

**Comment by FernandoMiguel** [comment]
I would do one container accessing multiple cross accounts via STS.
It's cleaner.
A much smaller and simpler cloudformation stack can create the role with the required priveleges and trust from the account/role running the scan, and can be destroyed later if need be.
To run a docker on every account, there will be far more modifications, VPC, SG, logs, etc

**Comment by 0xdabbad00** [comment]
Getting CloudMapper into Fargate is related to #287 for dockerizing it.

**Comment by 0xdabbad00** [comment]
@FernandoMiguel My thoughts on having one CloudMapper collect + audit container per account is that for an environment with 100's of accounts you would want CloudMapper to run in parallel, collecting data and auditing all of them simultaneously.  Otherwise you potentially bump into the problem of CloudMapper running in a series where each collect, cumulatively takes over 24 hours which would cause problems.

An alternative would be to make CloudMapper (or how it is run within the container) be able to run multiple collects simultaneously, which seems harder.

**Comment by FernandoMiguel** [comment]
depending on the size of the account and time it takes, you can run one docker task per account
or one task per service per account.
they do something, they die.. until everything is covered

**Comment by 0xdabbad00** [comment]
Audit filtering added in #492

To update where things are at, I have yet to do the following:
- Dockerize CloudMapper (#287)
- Have a way to send alerts to Slack for the audit items it finds.
- Have a way to control the container so that when it starts it knows:
  - What S3 bucket to get its config.json from, and any other config info, such as the new `config/audit_config_override.yaml`, and store its account_data to.
  - What accounts to run `collect` against and then `audit`
  - Where to send the alerts

**Comment by 0xdabbad00** [comment]
CloudMapper can now run inside a Docker container, but it doesn't do anything. The use case for that creation was to be able to make that setup seemless primarily so it can be used interactively.

The Slack output was added in #495.

Now that remains is the work in actually automatically deploying this into Fargate with some way of telling it what to do.

**Comment by 0xdabbad00** [comment]
I have a private repo I'm working in on a CDK app for this.  It creates an ECS.  It will eventually:
- Have an IAM role that can assume a role named `CloudMapper` in any account (likely to be configurable).
- Have an S3 bucket configured for it, that it will know to grab a script from to run, and also to store the `account-data` files there.
- Receive the Slack webhook somehow (maybe through Secrets manager, although I don't really view the webhooks as that secret, as the worst thing someone can do with it is DoS your channel with non-sense as it is write-only).
- Have the ECS task run on a schedule from CloudWatch Events.
- Have the ECS write log messages to CloudWatch Logs
- Have a CloudWatch Alarm setup to alert on any errors from CloudMapper

**Comment by 0xdabbad00** [comment]
I'm somewhat following https://github.com/aws-samples/aws-cdk-changelogs-demo/blob/master/changelogs-md.js?source=post_page---------------------------#L214

**Comment by 0xdabbad00** [comment]
This is complete and it is glorious! I have the code in a private repo as I let it run in some environments and work through the best way of making this public.

**Comment by 0xdabbad00** [comment]
I successfully deployed this in another environment and worked out some further bugs with it. Deployment generally was very smooth though.  Just waiting on some decisions to be made now on how to release it.

**Comment by 0xdabbad00** [comment]
I got the green light on open-sourcing this by doing so within this repo as opposed to a new one.  Probably mid-September that will happen.

**Comment by 0xdabbad00** [comment]
My understanding now is this should be published within two weeks.

**Comment by 0xdabbad00** [comment]
October 1 is the date.

**Comment by 0xdabbad00** [comment]
Resolved in #562. Blog post announcement to go out tomorrow morning.

**Comment by 0xdabbad00** [comment]
Blog post: https://duo.com/blog/continuous-auditing-with-cloudmapper
